File: tinkerpop3/src/test/java/overflowdb/testdomains/gratefuldead/SongDb.java
package overflowdb.testdomains.gratefuldead;

import overflowdb.NodeLayoutInformation;
import overflowdb.NodeRef;
import overflowdb.OdbNode;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

public class SongDb extends OdbNode {
  protected SongDb(NodeRef ref) {
    super(ref);
  }

  private String _name;
  private String _songType;
  private Integer _performances;

  public String name() {
    return _name;
  }

  public String songType() {
    return _songType;
  }

  public Integer performances() {
    return _performances;
  }

  @Override
  public NodeLayoutInformation layoutInformation() {
    return layoutInformation;
  }

  @Override
  public <A> A property2(String key) {
    if (Song.NAME.equals(key)) {
      return (A) _name;
    } else if (key == Song.SONG_TYPE) {
      return (A) _songType;
    } else if (key == Song.PERFORMANCES) {
      return (A) _performances;
    } else {
      return null;
    }
  }

  @Override
  public Map<String, Object> valueMap() {
    Map<String, Object> properties = new HashMap<>();
    if (_name != null) properties.put(Song.NAME, _name);
    if (_songType != null) properties.put(Song.SONG_TYPE, _songType);
    if (_performances != null) properties.put(Song.PERFORMANCES, _performances);
    return properties;
  }

  @Override
  protected void updateSpecificProperty(String key, Object value) {
    if (Song.NAME.equals(key)) {
      this._name = (String) value;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = (String) value;
    } else if (Song.PERFORMANCES.equals(key)) {
      this._performances = ((Integer) value);
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }


  @Override
  protected void removeSpecificProperty(String key) {
    if (Song.NAME.equals(key)) {
      this._name = null;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = null;
    } else if (Song.PERFORMANCES.equals(key)) {
      this._performances = null;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  public static NodeLayoutInformation layoutInformation = new NodeLayoutInformation(
      1,
      new HashSet<>(Arrays.asList(Song.NAME, Song.SONG_TYPE, Song.PERFORMANCES)),
      Arrays.asList(SungBy.layoutInformation, WrittenBy.layoutInformation, FollowedBy.layoutInformation),
      Arrays.asList(FollowedBy.layoutInformation));
}

File: traversal/src/test/scala/overflowdb/traversal/testdomains/gratefuldead/SongDb.scala
package overflowdb.traversal.testdomains.gratefuldead

import org.apache.tinkerpop.gremlin.structure.VertexProperty
import overflowdb.traversal._
import overflowdb.{NodeRef, OdbNode}

class SongDb(ref: NodeRef[SongDb]) extends OdbNode(ref) {
  private var _name: String = null
  private var _songType: String = null
  private var _performances: Integer = null

  def name: String = _name
  def songType: String = _songType
  def performances: Integer = _performances

  /* Song --- followedBy --- Song */
  def followedBy: Traversal[Song] = out(FollowedBy.Label).toScalaAs[Song]

  /* Artist <-- sungBy --- Song */
  def sungBy: Traversal[Artist] = out(SungBy.Label).toScalaAs[Artist]

  /* Artist <-- writtenBy --- Song */
  def writtenBy: Traversal[Artist] = out(WrittenBy.Label).toScalaAs[Artist]

  override def valueMap = {
    val properties = new java.util.HashMap[String, Any]
    if (_name != null) properties.put(Song.PropertyNames.Name, _name)
    if (_songType != null) properties.put(Song.PropertyNames.SongType, _songType)
    if (_performances != null) properties.put(Song.PropertyNames.Performances, _performances)
    properties
  }

  override protected def specificProperty2(key: String) =
    key match {
      case Song.PropertyNames.Name => _name
      case Song.PropertyNames.SongType => _songType
      case Song.PropertyNames.Performances => _performances
      case _ => null
    }

  override protected def updateSpecificProperty[V](cardinality: VertexProperty.Cardinality, key: String, value: V) =
    key match {
      case Song.PropertyNames.Name =>
        _name = value.asInstanceOf[String]
        property(Song.PropertyNames.Name)
      case Song.PropertyNames.SongType =>
        _songType = value.asInstanceOf[String]
        property(Song.PropertyNames.SongType)
      case Song.PropertyNames.Performances =>
        _performances = value.asInstanceOf[Integer]
        property(Song.PropertyNames.Performances)
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def removeSpecificProperty(key: String) =
    key match {
      case Song.PropertyNames.Name => _name = null
      case Song.PropertyNames.SongType => _songType = null
      case Song.PropertyNames.Performances => _performances = null
      case _ =>
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def layoutInformation = Song.layoutInformation
}
